Adds validation loss to LoRA fine tune single device

MaxFrax opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
https://github.com/pytorch/torchtune/issues/1042

#### Changelog
What are the changes made in this PR?
Adds support to a validation dataset and computes the loss on it after each epoch.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

MaxFrax commented:
@felipemello1 Finally I have been able to work on this. I'll make my way through the testing plan, but feel free to share any comment you might already have.

felipemello1 commented:
hey @MaxFrax , thank you! I am on PTO this week. I will get to it next week if someone doesnt do it before me.

ebsmothers reviewed (commented):
Hey @MaxFrax, thanks for this PR! The validation loop itself looks pretty reasonable, but I think we should figure out the right way to integrate it. E.g. right now it seems like we perform validation after every training epoch inside of the `train` method. Personally I would be in favor of splitting out into multiple methods to make things clearer. That will be a bit more work, but I want to make sure we expose this as clearly as possible. What about something like this?

```
def validate():
	# Should be roughly the code you added

def train():
	# Keep this mostly as it is, but add something like:
	if self.global_step % self.run_val_every_n_steps == 0:
		self.validate()
```

Then we can expose `run_val_every_n_steps` via config. A couple other things to think about would be a maximum number of batches in the val loop and early stopping. I don't think we need to worry about the latter for this PR, but should make sure it's something we're able to support later on.

Also cc @joecummings @felipemello1 @calvinpelletier for anything I've missed here.

ebsmothers commented on recipes/lora_finetune_single_device.py:
```diff
@@ -779,6 +788,38 @@ def train(self) -> None:
+                for idx, batch in enumerate(self._dataloader_val):
+                    utils.batch_to_device(batch, self._device)
+
+                    current_loss = self._loss_step(batch)
```
I think we will need to toggle eval <-> train mode for the model, right? (Another reason having a separate method will probably be cleanest)

MaxFrax replied:
Thanks @ebsmothers ! It makes sense. I will look into it. Do you have any pointers on how to do that?

ebsmothers commented on recipes/lora_finetune_single_device.py:
```diff
@@ -779,6 +788,38 @@ def train(self) -> None:
                 )
 
+                # VALIDATION LOOP
+                self._sampler_val.set_epoch(curr_epoch)
```
At least as is it's not clear to me why we want to set the epoch on the val sampler. If we're running over the entire val dataloader once why does the order matter? (And if we're not this will make the val loss noisier, right? Is that what we want?)

MaxFrax replied:
If running all over it, you are right, it doesn't make sense. When running only on a subset of the validation set, it might be interesting from a statistical perspective. In fact, testing with different random samples gives a more robust signal of proper generalization from the model. The argument against it is that it makes debugging difficult. For instance, you have some spike in the validation loss in your batch. Without random sampling, it’s easier to investigate it. Considering everything, I removed it.

MaxFrax commented:
Hi @ebsmothers ! I have updated the pr with the following edits, as per your recommendation:
- Created a stand alone method for the validation loop
- Parameter `run_val_every_n_steps` to invoke validate in specific points of the training epoch
- I also added `max_validation_batches` to cap the amount of batches run in each validation step

If there's any other feedback or comment, just let me know!

felipemello1 commented:
Thanks for making the changes! I will take a look at this PR later today.

RdoubleA reviewed (commented):
This looks good to me, although I'm getting worried about the proliferation of `cfg.get` and cfg validation logic in the recipes. There's nothing inherently wrong about the `cfg.get`, but it encourages the use of hidden parameters not exposed to the user. I don't have a good long term solution for this, but since we are only modifying one recipe, maybe we could at least update the lora single device configs to expose these fields so users know that it exists and we can check if the cfg field is None directly?
```
dataset_validation: null
run_val_every_n_steps: null
max_validation_batches: -1
```
I know this will affect a lot of files, so open to thoughts. We could also do this in a follow-up.

felipemello1 commented:
> maybe we could at least update the lora single device configs to expose these fields so users know that it exists and we can check if the cfg field is None directly?

Lets do this as a follow up. I can use my script to bulk update. But lets make sure that we all agree on how it should like in the config.

felipemello1 commented on recipes/lora_finetune_single_device.py:
```diff
@@ -652,6 +675,43 @@ def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
 
         return loss
 
+    def validate(self, curr_epoch) -> None:
```
1) Do we have model.eval() somewhere?

usually we want to set the model to .eval mode, because some layers have different behavior, like dropout.

By doing that, we then require less memory, because we only need the forward pass, which allows us to have a higher batch_size --> faster validation step.

I am not sure about the implications it may have to compile/FSDP. For example, compile will have to create a new graph that doesnt require grad, so compile time will have to increase. If the number of graph breaks increase, we may have to manually change the threshold of maximum number of graph breaks. (there is an example of that in one of our RL recipes)

2) not all recipes have self._loss_step. We would have to standardize and make sure that they all do, but this requires a different PR,.

IMO, if you have access to >1 GPU, I would encourage you to implement it in lora_distributed with QLoRA config, add .eval(), run it:
- with eval + compile + opt_in_bwd + activation ckpt + activation offloading
- without eval + compile + opt_in_bwd + activation ckpt + activation offloading

If nothing breaks, I would feel more confident in approving it

Ps: we would also have to add mode.train() in the training loop

MaxFrax replied:
@felipemello1 Thanks for the detailed breakdown and suggestions. Should we also unload the model being trained before loading the eval one? Having just one in memory would allow for bigger batch sizes.

That said, I’m currently constrained on time and not very familiar with the implementation details for this. If I were to take this on, it would likely take me a significant amount of time to get it done properly.

Would you be able to take the lead on this?

felipemello1 replied:
hey @MaxFrax , completely understandable. Thanks for sharing it.

I dont think that I will have bandwidth soon, but if i do, this PR is a good start. 

@Ankur-singh , pinging you in case you are looking for more issues to contribute to! :D 

Thank you guys!

felipemello1 commented on recipes/lora_finetune_single_device.py:
```diff
@@ -652,6 +675,43 @@ def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
+
+            # This bit allows to see the loss for each batch. Not sure about step indexing.
+            log_dict = {
+                "val_loss": current_loss.item(),
```
I wonder if we should be logging memory/TPS too. If memory is very low, this would show to the user that they can increase bsz. What do you think?

MaxFrax replied:
It makes sense. I will definitely look into it.
By the way, when is the previous training batch deallocated from memory? Do I have to deallocate manually? It would be handy to do so before staring the validation step to have more memory available.

felipemello1 commented on recipes/lora_finetune_single_device.py:
```diff
@@ -779,6 +839,12 @@ def train(self) -> None:
 
+                if (
+                    self.run_validation
+                    and self.global_step % self.run_val_every_n_steps == 0
```
i think it makes a lot of sense to eval every N steps, but currently a lot of our training logic is based on epochs. I wonder if we should honor this and keep it based on epoch. Maybe users could pass a float, e.g. every 0.5 epochs.

Not 100% sure about this, just brainstorming.

MaxFrax replied:
I'm happy to change it as you suggest, I like it better too. I just want to point out that the scheduler has the parameter `num_warmup_steps` which contradicts your statement:
> our training logic is based on epochs

As user, I'd love the num_warmup_steps to be based on epochs as well.

felipemello1 reviewed (commented):
thanks for the PR! It looks simple and the functions make sense!

I added a few comments/ideas. Please push back on what you disagree. 

IMO, to approve this, we would need two things:

1. Testing, like i suggested in one of the comments. Let me know if you are comfortable running it, otherwise we can help you out
2. An example of how the config should look like. The UI should play a big factor on this PR.

felipemello1 commented:
@MaxFrax

> Should I provide a recipe using the validation dataset? Are we talking about the docs?
Let me know more precisely what I should do, and I'll be happy to look into it.

Your PR only contains changes to the recipe. I would encourage you to:
- make changes to one of the configs too, to illustrate how users would use it
- Put the command in the description of the PR
- share an image of the logs generated in weights and biases under the testing section

MaxFrax commented:
> @MaxFrax
> 
> > Should I provide a recipe using the validation dataset? Are we talking about the docs?
> > Let me know more precisely what I should do, and I'll be happy to look into it.
> 
> Your PR only contains changes to the recipe. I would encourage you to:
> 
> * make changes to one of the configs too, to illustrate how users would use it
> * Put the command to launch this config in the description of the PR, under the testing section
> * share an image of the logs generated in weights and biases under the testing section

Thanks for the feedback!

I have edited the llama3_2/1B_lora_single_device config and uploaded the W&B screen (I capped the training steps and changed batch sizes to make the screenshot instead of running exactly the committed recipe).
I’ve resolved all the issues raised in the PR review, except for the model evaluation mode and the validation memory logging.
